core/roslib/src/roslib/packages.py

```diff
@@ -509,3 +515,16 @@ def find_resource(pkg, resource_name, filter_fn=None, rospack=None):
         if match not in unique_matches:
             unique_matches.append(match)
     return unique_matches
+
+def catkin_packages_cache():
+    """
+    Return packages which can be used in source_path_to_packages keyword
+    of catkin.find_in_workspaces.find_in_workspaces
+    """
+    source_path_to_packages = {}
+    workspaces = catkin_get_workspaces()
+    for workspace in workspaces:
+        source_spaces = catkin_get_source_paths(workspace)
+        for source_path in source_spaces:
+            source_path_to_packages[source_path] = catkin_find_packages(source_path)
+    return source_path_to_packages
```

Comment:
The logic to iterate over the workspaces and finding packages in each of them is provided by catkin.find_in_workspaces and should be reused instead of being duplicated here.